mirgecom/viscous.py

```diff
@@ -275,63 +277,170 @@ def viscous_flux(state, grad_cv, grad_t):
             momentum=tau, species_mass=-j)
 
 
-def viscous_facial_flux(discr, state_tpair, grad_cv_tpair, grad_t_tpair,
-                        local=False):
-    """Return the viscous flux across a face given the solution on both sides.
+def viscous_divergence_flux(discr, state_pair, grad_cv_pair, grad_t_pair,
+                         beta=0., gamma=0, **kwargs):
+    r"""Return a viscous facial flux for the divergence operator.
+
+    The flux is defined as:
+
+    .. math::
+
+        f_{\text{face}} = \frac{1}{2}\left(\mathbf{f}_v^+
+        + \mathbf{f}_v^-\right)\cdot\hat{\mathbf{n}}
+        + \frac{\beta}{2}\left(\mathbf{f}_v^+ - \mathbf{f}_v^-\right),
+
+    with viscous fluxes ($\mathbf{f}_v$), and the outward pointing
+    face normal ($\hat{\mathbf{n}}$).
 
     Parameters
     ----------
     discr: :class:`~grudge.eager.EagerDGDiscretization`
 
         The discretization to use
 
-    state_tpair: :class:`~grudge.trace_pair.TracePair`
+    state_pair: :class:`~grudge.trace_pair.TracePair`
 
         Trace pair of :class:`~mirgecom.gas_model.FluidState` with the full fluid
         conserved and thermal state on the faces
 
-    grad_cv_tpair: :class:`~grudge.trace_pair.TracePair`
+    grad_cv_pair: :class:`~grudge.trace_pair.TracePair`
 
         Trace pair of :class:`~mirgecom.fluid.ConservedVars` with the gradient of the
         fluid solution on the faces
 
-    grad_t_tpair: :class:`~grudge.trace_pair.TracePair`
+    grad_t_pair: :class:`~grudge.trace_pair.TracePair`
 
         Trace pair of temperature gradient on the faces.
 
-    local: bool
-
-        Indicates whether to skip projection of fluxes to "all_faces" or not. If
-        set to *False* (the default), the returned fluxes are projected to
-        "all_faces".  If set to *True*, the returned fluxes are not projected to
-        "all_faces"; remaining instead on the boundary restriction.
-
     Returns
     -------
     :class:`~mirgecom.fluid.ConservedVars`
 
         The viscous transport flux in the face-normal direction on "all_faces" or
         local to the sub-discretization depending on *local* input parameter
     """
-    actx = state_tpair.int.array_context
-    dd = state_tpair.dd
-    dd_all_faces = dd.with_dtag("all_faces")
+    actx = state_pair.int.array_context
+    normal = thaw(discr.normal(state_pair.dd), actx)
+
+    f_int = viscous_flux(state_pair.int, grad_cv_pair.int,
+                         grad_t_pair.int)
+    f_ext = viscous_flux(state_pair.ext, grad_cv_pair.ext,
+                         grad_t_pair.ext)
+    f_pair = TracePair(state_pair.dd, interior=f_int, exterior=f_ext)
+    q_pair = TracePair(state_pair.dd, interior=state_pair.int.cv,
+                       exterior=state_pair.ext.cv)
+    from arraycontext import outer
+    jump_term = -gamma*outer(q_pair.diff, normal)/2
+
+    return divergence_flux(trace_pair=f_pair, normal=normal,
+                           alpha=jump_term, beta=beta)
```

Comment:
Ah, now I see how `alpha` is being used. Not really a fan of having an extra catch-all parameter to `divergence_flux` like that. If the form of the flux is significantly different from a standard central flux, it should probably be its own separate flux function.

Also, if the intent of `gamma` is to support an interior penalty method, keep in mind that the form of the penalty coefficient may be complicated and not really something you want to expose in the public viscous flux interface. Might be best to leave this out for now and add a separate `viscous_facial_flux_penalty` or something later.

Reply:
These were just based on the fluxes Mengaldo used.  With the default parameters, they end up being central like we have been using.

Reply:
I think my main issues with these are:

1. The names are too general (LFR and HLL are also divergence fluxes, after all)
2. The `alpha` parameter in `divergence_flux` is too open-ended.

If we moved the `jump_term` stuff inside the divergence flux function and found some better names, I think I would be OK with them.

What do you think about integrating these functions into the lower-level flux interface so they're similar to `num_flux_hll`, etc.? Something like https://github.com/majosm/mirgecom/commit/22d1af3714b20f0a468c1d403169a649bbecfe50.

Reply:
Sure this is just fine with me.

Reply:
Actually, can we please keep the interface naming such that it is clear what operator (divergence) vs. (gradient) the flux is intended for?   For the former the return is a scalar for each eqn and for the latter it is a vector quantity for each eqn.

Reply:
ok, let's do that when we add new fluxes.  for now i just removed all the fluxes that you having a problem with.  can we push on without them for now?